catroid/src/main/java/org/catrobat/catroid/ui/MultilingualActivity.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.ui;
+
+import android.annotation.TargetApi;
+import android.app.Activity;
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.content.res.Configuration;
+import android.content.res.Resources;
+import android.os.Build;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.util.DisplayMetrics;
+import android.view.View;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.ListView;
+
+import org.catrobat.catroid.R;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+import static org.catrobat.catroid.CatroidApplication.defaultSystemLanguage;
+import static org.catrobat.catroid.CatroidApplication.languageSharedPreferences;
+import static org.catrobat.catroid.common.Constants.LANGUAGE_CODE;
+import static org.catrobat.catroid.common.Constants.LANGUAGE_TAG_KEY;
+
+public class MultilingualActivity extends Activity {
```

Comment:
As far as i got it, this is the activity to set the language right?

Reply:
right

Reply:
Ok I looked a little further into this.

As you saw, we are have our `SettingsActivity` which is derrived from Android's `PreferencesActivity`.
Now, `PreferencesActivity` is a really easy way to set up preference menus which can have different screens and selection options.

If you look into the `preferences.xml` you see a good example, look for `setting_mindstorms_nxt_sensor_1`. This is a `ListPreference` meaning you are able to choose a setting (in your case a language) from an array. 
The array values and display names are specified in `arrays.xml`.

This `ListPreference` is nested within a `PreferenceScreen` in the `preferences.xml`. Meaning it automatically switches screens (which looks just the way like it would be in a new Activity).

Now this has multiple benefits:
First of all it is really easy to implement, and the Android system takes care of all the screen switching and settings value setting and creating the layout for you.
Most importantly the Activity stack stays clean. Meaning: the way you implemented it right now pushes a new Activity on top of the stack (with all the inflating, creating, filling of the UI etc). If not properly handled, this can mess up things pretty bad. For example, if you open the language setting now and press back, you are suddenly way back at the Main Menu. In the other preference screens you always get one level up, to where you came from.
As @84n4n4 pointed out to me, this is particularly bad considering tests (which really should avoid activity switching) in a way that it may run successful on this test, but somehow make tests executed later in the test run fail randomly.
This is close to impossible to debug and must be avoided by all means.

Reply:
ok, I will refactor it as you explained, as I see, should the test be written in SettingsActivityTest. right?

Reply:
without activity switching, how will be the test? I have no idea :(